File: 3.x.x/crates-utopia/query/src/tokenizer.rs
//! PHP `Utopia\Query\Tokenizer`.

use crate::error::QueryError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TokenType {
    Keyword,
    Identifier,
    QuotedIdentifier,
    Integer,
    Float,
    String,
    Boolean,
    Null,
    Operator,
    LeftParen,
    RightParen,
    Comma,
    Semicolon,
    Dot,
    Star,
    Placeholder,
    NamedPlaceholder,
    NumberedPlaceholder,
    LineComment,
    BlockComment,
    Whitespace,
    Eof,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub token_type: TokenType,
    pub value: String,
    pub position: usize,
}

impl Token {
    pub fn new(token_type: TokenType, value: impl Into<String>, position: usize) -> Self {
        Self {
            token_type,
            value: value.into(),
            position,
        }
    }
}

const KEYWORDS: &[&str] = &[
    "SELECT",
    "FROM",
    "WHERE",
    "AND",
    "OR",
    "NOT",
    "JOIN",
    "LEFT",
    "RIGHT",
    "INNER",
    "OUTER",
    "FULL",
    "CROSS",
    "NATURAL",
    "ON",
    "AS",
    "ORDER",
    "BY",
    "GROUP",
    "HAVING",
    "LIMIT",
    "OFFSET",
    "ASC",
    "DESC",
    "IN",
    "BETWEEN",
    "LIKE",
    "ILIKE",
    "IS",
    "CASE",
    "WHEN",
    "THEN",
    "ELSE",
    "END",
    "EXISTS",
    "DISTINCT",
    "ALL",
    "UNION",
    "INTERSECT",
    "EXCEPT",
    "WITH",
    "RECURSIVE",
    "SET",
    "INSERT",
    "INTO",
    "VALUES",
    "UPDATE",
    "DELETE",
    "CREATE",
    "ALTER",
    "DROP",
    "TABLE",
    "INDEX",
    "VIEW",
    "OVER",
    "PARTITION",
    "WINDOW",
    "ROWS",
    "RANGE",
    "UNBOUNDED",
    "PRECEDING",
    "FOLLOWING",
    "CURRENT",
    "ROW",
    "FETCH",
    "NEXT",
    "FIRST",
    "LAST",
    "NULLS",
    "CAST",
    "FILTER",
    "WITHIN",
];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenizerKind {
    Generic,
    Mysql,
    Mariadb,
    Postgres,
    Sqlite,
    Clickhouse,
}

#[derive(Debug, Clone)]
pub struct Tokenizer {
    pub kind: TokenizerKind,
    sql: String,
    length: usize,
    pos: usize,
}

impl Tokenizer {
    pub fn new() -> Self {
        Self {
            kind: TokenizerKind::Generic,
            sql: String::new(),
            length: 0,
            pos: 0,
        }
    }

    pub fn mysql() -> Self {
        Self {
            kind: TokenizerKind::Mysql,
            ..Self::new()
        }
    }

    pub fn mariadb() -> Self {
        Self {
            kind: TokenizerKind::Mariadb,
            ..Self::new()
        }
    }

    pub fn postgres() -> Self {
        Self {
            kind: TokenizerKind::Postgres,
            ..Self::new()
        }
    }

    pub fn sqlite() -> Self {
        Self {
            kind: TokenizerKind::Sqlite,
            ..Self::new()
        }
    }

    pub fn clickhouse() -> Self {
        Self {
            kind: TokenizerKind::Clickhouse,
            ..Self::new()
        }
    }

    fn identifier_quote_char(&self) -> char {
        match self.kind {
            TokenizerKind::Postgres | TokenizerKind::Sqlite => '"',
            _ => '`',
        }
    }

    pub fn tokenize(&mut self, sql: &str) -> Result<Vec<Token>, QueryError> {
        let prepared = match self.kind {
            TokenizerKind::Mysql | TokenizerKind::Mariadb => replace_hash_comments(sql),
            _ => sql.to_owned(),
        };
        self.sql = prepared;
        self.length = self.sql.len();
        self.pos = 0;
        let mut tokens = Vec::new();
        let quote_char = self.identifier_quote_char();
        while self.pos < self.length {
            let start = self.pos;
            let char = self.byte_at(self.pos);
            let token = match char {
                b' ' | b'\t' | b'\n' | b'\r' => self.read_whitespace(start),
                b'-' => self.read_dash_prefix(start)?,
                b'/' => self.read_slash_prefix(start)?,
                b'\'' => self.read_string(start)?,
                q if q == quote_char as u8 => self.read_quoted_identifier(start, quote_char)?,
                b'"' => self.read_quoted_identifier(start, '"')?,
                b'0'..=b'9' => self.read_number(start),
                b'a'..=b'z' | b'A'..=b'Z' | b'_' => self.read_identifier_or_keyword(start),
                b'(' => self.consume_single(TokenType::LeftParen, "(", start),
                b')' => self.consume_single(TokenType::RightParen, ")", start),
                b',' => self.consume_single(TokenType::Comma, ",", start),
                b';' => self.consume_single(TokenType::Semicolon, ";", start),
                b'.' => self.read_dot(start),
                b'*' => self.consume_single(TokenType::Star, "*", start),
                b'?' => self.read_placeholder(start),
                b':' => self.read_named_placeholder(start),
                b'$' if self.kind == TokenizerKind::Postgres => {
                    self.read_numbered_placeholder(start)
                }
                _ => self.read_operator(start)?,
            };
            tokens.push(token);
        }
        tokens.push(Token::new(TokenType::Eof, "", self.pos));
        Ok(tokens)
    }

    pub fn filter(tokens: Vec<Token>) -> Vec<Token> {
        tokens
            .into_iter()
            .filter(|t| {
                !matches!(
                    t.token_type,
                    TokenType::Whitespace | TokenType::LineComment | TokenType::BlockComment
                )
            })
            .collect()
    }

    fn byte_at(&self, i: usize) -> u8 {
        self.sql.as_bytes().get(i).copied().unwrap_or(0)
    }

    fn consume_single(&mut self, ty: TokenType, value: &str, start: usize) -> Token {
        self.pos += 1;
        Token::new(ty, value, start)
    }

    fn read_whitespace(&mut self, start: usize) -> Token {
        while self.pos < self.length {
            match self.byte_at(self.pos) {
                b' ' | b'\t' | b'\n' | b'\r' => self.pos += 1,
                _ => break,
            }
        }
        Token::new(TokenType::Whitespace, &self.sql[start..self.pos], start)
    }

    fn read_dash_prefix(&mut self, start: usize) -> Result<Token, QueryError> {
        if self.pos + 1 < self.length && self.byte_at(self.pos + 1) == b'-' {
            self.pos += 2;
            while self.pos < self.length && self.byte_at(self.pos) != b'\n' {
                self.pos += 1;
            }
            return Ok(Token::new(
                TokenType::LineComment,
                &self.sql[start..self.pos],
                start,
            ));
        }
        self.read_operator(start)
    }

    fn read_slash_prefix(&mut self, start: usize) -> Result<Token, QueryError> {
        if self.pos + 1 < self.length && self.byte_at(self.pos + 1) == b'*' {
            self.pos += 2;
            while self.pos + 1 < self.length
                && !(self.byte_at(self.pos) == b'*' && self.byte_at(self.pos + 1) == b'/')
            {
                self.pos += 1;
            }
            if self.pos + 1 < self.length {
                self.pos += 2;
            }
            return Ok(Token::new(
                TokenType::BlockComment,
                &self.sql[start..self.pos],
                start,
            ));
        }
        self.read_operator(start)
    }

    fn read_string(&mut self, start: usize) -> Result<Token, QueryError> {
        self.pos += 1;
        while self.pos < self.length {
            let c = self.byte_at(self.pos);
            if c == b'\\' {
                self.pos += 2;
                continue;
            }
            if c == b'\'' {
                self.pos += 1;
                if self.pos < self.length && self.byte_at(self.pos) == b'\'' {
                    self.pos += 1;
                    continue;
                }
                let raw = &self.sql[start + 1..self.pos - 1];
                let value = raw.replace("''", "'").replace("\\'", "'");
                return Ok(Token::new(TokenType::String, value, start));
            }
            self.pos += 1;
        }
        Err(QueryError::validation("Unterminated string literal"))
    }

    fn read_quoted_identifier(&mut self, start: usize, quote: char) -> Result<Token, QueryError> {
        let q = quote as u8;
        self.pos += 1;
        while self.pos < self.length {
            let c = self.byte_at(self.pos);
            if c == q {
                self.pos += 1;
                if self.pos < self.length && self.byte_at(self.pos) == q {
                    self.pos += 1;
                    continue;
                }
                let raw = &self.sql[start + 1..self.pos - 1];
                let doubled = format!("{quote}{quote}");
                let value = raw.replace(&doubled, &quote.to_string());
                return Ok(Token::new(TokenType::QuotedIdentifier, value, start));
            }
            self.pos += 1;
        }
        Err(QueryError::validation("Unterminated quoted identifier"))
    }

    fn read_number(&mut self, start: usize) -> Token {
        let mut is_float = false;
        while self.pos < self.length {
            match self.byte_at(self.pos) {
                b'0'..=b'9' => self.pos += 1,
                b'.' if !is_float => {
                    is_float = true;
                    self.pos += 1;
                }
                b'e' | b'E' => {
                    is_float = true;
                    self.pos += 1;
                    if self.pos < self.length
                        && (self.byte_at(self.pos) == b'+' || self.byte_at(self.pos) == b'-')
                    {
                        self.pos += 1;
                    }
                }
                _ => break,
            }
        }
        Token::new(
            if is_float {
                TokenType::Float
            } else {
                TokenType::Integer
            },
            &self.sql[start..self.pos],
            start,
        )
    }

    fn read_identifier_or_keyword(&mut self, start: usize) -> Token {
        while self.pos < self.length {
            match self.byte_at(self.pos) {
                b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' => self.pos += 1,
                _ => break,
            }
        }
        let value = self.sql[start..self.pos].to_owned();
        let upper = value.to_ascii_uppercase();
        if upper == "TRUE" || upper == "FALSE" {
            return Token::new(TokenType::Boolean, value, start);
        }
        if upper == "NULL" {
            return Token::new(TokenType::Null, value, start);
        }
        if KEYWORDS.contains(&upper.as_str()) {
            Token::new(TokenType::Keyword, upper, start)
        } else {
            Token::new(TokenType::Identifier, value, start)
        }
    }

    fn read_dot(&mut self, start: usize) -> Token {
        if self.pos + 1 < self.length && self.byte_at(self.pos + 1).is_ascii_digit() {
            return self.read_number(start);
        }
        self.consume_single(TokenType::Dot, ".", start)
    }

    fn read_placeholder(&mut self, start: usize) -> Token {
        self.pos += 1;
        if self.pos < self.length && self.byte_at(self.pos).is_ascii_digit() {
            while self.pos < self.length && self.byte_at(self.pos).is_ascii_digit() {
                self.pos += 1;
            }
            return Token::new(
                TokenType::NumberedPlaceholder,
                &self.sql[start..self.pos],
                start,
            );
        }
        Token::new(TokenType::Placeholder, "?", start)
    }

    fn read_named_placeholder(&mut self, start: usize) -> Token {
        self.pos += 1;
        while self.pos < self.length {
            match self.byte_at(self.pos) {
                b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' => self.pos += 1,
                _ => break,
            }
        }
        Token::new(
            TokenType::NamedPlaceholder,
            &self.sql[start..self.pos],
            start,
        )
    }

    fn read_numbered_placeholder(&mut self, start: usize) -> Token {
        self.pos += 1;
        while self.pos < self.length && self.byte_at(self.pos).is_ascii_digit() {
            self.pos += 1;
        }
        Token::new(
            TokenType::NumberedPlaceholder,
            &self.sql[start..self.pos],
            start,
        )
    }

    fn read_operator(&mut self, start: usize) -> Result<Token, QueryError> {
        let rest = &self.sql[self.pos..];
        let two = if rest.len() >= 2 { &rest[..2] } else { "" };
        let three = if rest.len() >= 3 { &rest[..3] } else { "" };
        let (op, len) = if matches!(three, "<=>" | "<<>" | "<>>") {
            (three, 3)
        } else if matches!(
            two,
            "<=" | ">=" | "<>" | "!=" | "||" | "&&" | "->" | ">>" | "<<" | ":=" | "::"
        ) {
            (two, 2)
        } else if rest.starts_with("->>") {
            ("->>", 3)
        } else if !rest.is_empty()
            && matches!(
                rest.as_bytes()[0],
                b'=' | b'<' | b'>' | b'+' | b'-' | b'/' | b'%' | b'!' | b'|' | b'&' | b'~' | b'^'
            )
        {
            (&rest[..1], 1)
        } else {
            return Err(QueryError::validation(format!(
                "Unexpected character at position {start}"
            )));
        };
        self.pos += len;
        Ok(Token::new(TokenType::Operator, op, start))
    }
}

impl Default for Tokenizer {
    fn default() -> Self {
        Self::new()
    }
}

fn replace_hash_comments(sql: &str) -> String {
    let bytes = sql.as_bytes();
    let mut result = String::new();
    let mut i = 0;
    while i < bytes.len() {
        let char = bytes[i];
        if char == b'\'' {
            result.push('\'');
            i += 1;
            while i < bytes.len() {
                let c = bytes[i];
                if c == b'\\' {
                    result.push('\\');
                    i += 1;
                    if i < bytes.len() {
                        result.push(bytes[i] as char);
                        i += 1;
                    }
                    continue;
                }
                if c == b'\'' {
                    result.push('\'');
                    i += 1;
                    if i < bytes.len() && bytes[i] == b'\'' {
                        result.push('\'');
                        i += 1;
                        continue;
                    }
                    break;
                }
                result.push(c as char);
                i += 1;
            }
            continue;
        }
        if char == b'`' || char == b'"' {
            let q = char as char;
            result.push(q);
            i += 1;
            while i < bytes.len() {
                let c = bytes[i];
                if c == char {
                    result.push(q);
                    i += 1;
                    if i < bytes.len() && bytes[i] == char {
                        result.push(q);
                        i += 1;
                        continue;
                    }
                    break;
                }
                result.push(c as char);
                i += 1;
            }
            continue;
        }
        if char == b'#' {
            result.push_str("--");
            i += 1;
            continue;
        }
        result.push(char as char);
        i += 1;
    }
    result
}

pub type MySQL = Tokenizer;
pub type MariaDB = Tokenizer;
pub type PostgreSQL = Tokenizer;
pub type SQLite = Tokenizer;
pub type ClickHouse = Tokenizer;

File: packages/query/src/Query/Tokenizer/Tokenizer.php
<?php

namespace Utopia\Query\Tokenizer;

use Utopia\Query\Exception\ValidationException;

class Tokenizer
{
    private const KEYWORD_MAP = [
        'SELECT' => true, 'FROM' => true, 'WHERE' => true, 'AND' => true,
        'OR' => true, 'NOT' => true, 'JOIN' => true, 'LEFT' => true,
        'RIGHT' => true, 'INNER' => true, 'OUTER' => true, 'FULL' => true,
        'CROSS' => true, 'NATURAL' => true, 'ON' => true, 'AS' => true,
        'ORDER' => true, 'BY' => true, 'GROUP' => true, 'HAVING' => true,
        'LIMIT' => true, 'OFFSET' => true, 'ASC' => true, 'DESC' => true,
        'IN' => true, 'BETWEEN' => true, 'LIKE' => true, 'ILIKE' => true,
        'IS' => true, 'CASE' => true, 'WHEN' => true, 'THEN' => true,
        'ELSE' => true, 'END' => true, 'EXISTS' => true, 'DISTINCT' => true,
        'ALL' => true, 'UNION' => true, 'INTERSECT' => true, 'EXCEPT' => true,
        'WITH' => true, 'RECURSIVE' => true, 'SET' => true, 'INSERT' => true,
        'INTO' => true, 'VALUES' => true, 'UPDATE' => true, 'DELETE' => true,
        'CREATE' => true, 'ALTER' => true, 'DROP' => true, 'TABLE' => true,
        'INDEX' => true, 'VIEW' => true, 'OVER' => true, 'PARTITION' => true,
        'WINDOW' => true, 'ROWS' => true, 'RANGE' => true, 'UNBOUNDED' => true,
        'PRECEDING' => true, 'FOLLOWING' => true, 'CURRENT' => true,
        'ROW' => true, 'FETCH' => true, 'NEXT' => true, 'FIRST' => true,
        'LAST' => true, 'NULLS' => true, 'CAST' => true, 'FILTER' => true,
        'WITHIN' => true,
    ];

    /**
     * Single-character operator lookup table. Used by tryReadOperator to
     * avoid allocating a haystack array on every character.
     */
    private const SINGLE_OPERATORS = [
        '=' => true,
        '<' => true,
        '>' => true,
        '+' => true,
        '-' => true,
        '/' => true,
        '%' => true,
    ];

    private string $sql;

    private int $length;

    private int $pos;

    /**
     * @return Token[]
     */
    public function tokenize(string $sql): array
    {
        $this->sql = $sql;
        $this->length = strlen($sql);
        $this->pos = 0;

        $tokens = [];
        $quoteChar = $this->getIdentifierQuoteChar();

        while ($this->pos < $this->length) {
            $start = $this->pos;
            $char = $this->sql[$this->pos];

            $tokens[] = match (true) {
                $char === ' ' || $char === "\t" || $char === "\n" || $char === "\r" => $this->readWhitespace($start),
                $char === '-' => $this->readDashPrefix($start),
                $char === '/' => $this->readSlashPrefix($start),
                $char === '\'' => $this->readString($start),
                $char === $quoteChar => $this->readQuotedIdentifier($start, $quoteChar),
                $char === '"' => $this->readQuotedIdentifier($start, '"'),
                $char >= '0' && $char <= '9' => $this->readNumber($start),
                ($char >= 'a' && $char <= 'z') || ($char >= 'A' && $char <= 'Z') || $char === '_' => $this->readIdentifierOrKeyword($start),
                $char === '(' => $this->consumeSingleChar(TokenType::LeftParen, '(', $start),
                $char === ')' => $this->consumeSingleChar(TokenType::RightParen, ')', $start),
                $char === ',' => $this->consumeSingleChar(TokenType::Comma, ',', $start),
                $char === ';' => $this->consumeSingleChar(TokenType::Semicolon, ';', $start),
                $char === '.' => $this->readDot($start),
                $char === '*' => $this->consumeSingleChar(TokenType::Star, '*', $start),
                $char === '?' => $this->consumeSingleChar(TokenType::Placeholder, '?', $start),
                $char === ':' => $this->readColonPrefix($start),
                $char === '$' => $this->readDollarPrefix($start),
                default => $this->readOperatorOrUnknown($start, $char),
            };
        }

        $tokens[] = new Token(TokenType::Eof, '', $this->pos);

        return $tokens;
    }

    private function consumeSingleChar(TokenType $type, string $value, int $start): Token
    {
        $this->pos++;
        return new Token($type, $value, $start);
    }

    private function readDashPrefix(int $start): Token
    {
        if ($this->peek(1) === '-') {
            return $this->readLineComment($start);
        }

        return $this->readOperatorOrUnknown($start, '-');
    }

    private function readSlashPrefix(int $start): Token
    {
        if ($this->peek(1) === '*') {
            return $this->readBlockComment($start);
        }

        return $this->readOperatorOrUnknown($start, '/');
    }

    private function readDot(int $start): Token
    {
        $next = $this->peek(1);
        if ($next !== null && $this->isDigit($next)) {
            return $this->readNumber($start);
        }

        $this->pos++;
        return new Token(TokenType::Dot, '.', $start);
    }

    private function readColonPrefix(int $start): Token
    {
        $next = $this->peek(1);
        if ($next === ':') {
            $this->pos += 2;
            return new Token(TokenType::Operator, '::', $start);
        }
        if ($next !== null && $this->isIdentStart($next)) {
            return $this->readNamedPlaceholder($start);
        }

        $this->pos++;
        return new Token(TokenType::Operator, ':', $start);
    }

    private function readDollarPrefix(int $start): Token
    {
        $next = $this->peek(1);
        if ($next !== null && $this->isDigit($next)) {
            return $this->readNumberedPlaceholder($start);
        }

        $this->pos++;
        return new Token(TokenType::Operator, '$', $start);
    }

    private function readOperatorOrUnknown(int $start, string $char): Token
    {
        $op = $this->tryReadOperator($start);
        if ($op !== null) {
            return $op;
        }

        // Emit unknown characters as single-char operator tokens
        $this->pos++;
        return new Token(TokenType::Operator, $char, $start);
    }

    /**
     * @param Token[] $tokens
     * @return Token[]
     */
    public static function filter(array $tokens): array
    {
        $result = [];
        foreach ($tokens as $token) {
            if (
                $token->type !== TokenType::Whitespace
                && $token->type !== TokenType::LineComment
                && $token->type !== TokenType::BlockComment
            ) {
                $result[] = $token;
            }
        }
        return $result;
    }

    protected function getIdentifierQuoteChar(): string
    {
        return '`';
    }

    private function peek(int $offset): ?string
    {
        $idx = $this->pos + $offset;
        return $idx < $this->length ? $this->sql[$idx] : null;
    }

    private function isDigit(string $char): bool
    {
        return $char >= '0' && $char <= '9';
    }

    private function isIdentStart(string $char): bool
    {
        return ($char >= 'a' && $char <= 'z')
            || ($char >= 'A' && $char <= 'Z')
            || $char === '_';
    }

    private function isIdentChar(string $char): bool
    {
        return $this->isIdentStart($char) || $this->isDigit($char);
    }

    private function readWhitespace(int $start): Token
    {
        while ($this->pos < $this->length) {
            $c = $this->sql[$this->pos];
            if ($c === ' ' || $c === "\t" || $c === "\n" || $c === "\r") {
                $this->pos++;
            } else {
                break;
            }
        }

        return new Token(TokenType::Whitespace, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readLineComment(int $start): Token
    {
        $this->pos += 2;
        while ($this->pos < $this->length && $this->sql[$this->pos] !== "\n") {
            $this->pos++;
        }

        return new Token(TokenType::LineComment, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readBlockComment(int $start): Token
    {
        $this->pos += 2;
        $terminated = false;
        while ($this->pos < $this->length - 1) {
            if ($this->sql[$this->pos] === '*' && $this->sql[$this->pos + 1] === '/') {
                $this->pos += 2;
                $terminated = true;
                break;
            }
            $this->pos++;
        }

        if (!$terminated) {
            $this->pos = $this->length;
        }

        return new Token(TokenType::BlockComment, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readString(int $start): Token
    {
        $this->pos++;
        $terminated = false;
        while ($this->pos < $this->length) {
            $char = $this->sql[$this->pos];
            if ($char === '\\') {
                // Backslash escape: skip next character. Reject trailing backslash at EOF.
                if ($this->pos + 1 >= $this->length) {
                    throw new ValidationException('Unterminated string literal');
                }
                $this->pos += 2;
                continue;
            }
            if ($char === '\'') {
                $this->pos++;
                // Check for escaped quote ''
                if ($this->pos < $this->length && $this->sql[$this->pos] === '\'') {
                    $this->pos++;
                    continue;
                }
                $terminated = true;
                break;
            }
            $this->pos++;
        }

        if (!$terminated) {
            throw new ValidationException('Unterminated string literal');
        }

        return new Token(TokenType::String, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readQuotedIdentifier(int $start, string $quote): Token
    {
        $this->pos++;
        $terminated = false;
        while ($this->pos < $this->length) {
            if ($this->sql[$this->pos] === $quote) {
                $this->pos++;
                // Check for escaped quote (doubled)
                if ($this->pos < $this->length && $this->sql[$this->pos] === $quote) {
                    $this->pos++;
                    continue;
                }
                $terminated = true;
                break;
            }
            $this->pos++;
        }

        if (!$terminated) {
            throw new ValidationException('Unterminated quoted identifier');
        }

        return new Token(TokenType::QuotedIdentifier, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readNumber(int $start): Token
    {
        $isFloat = false;

        // Handle dot-prefixed floats like .5
        if ($this->pos < $this->length && $this->sql[$this->pos] === '.') {
            $isFloat = true;
            $this->pos++;
        }

        while ($this->pos < $this->length && $this->isDigit($this->sql[$this->pos])) {
            $this->pos++;
        }

        if (!$isFloat && $this->pos < $this->length && $this->sql[$this->pos] === '.') {
            $next = $this->peek(1);
            if ($next !== null && $this->isDigit($next)) {
                $isFloat = true;
                $this->pos++;
                while ($this->pos < $this->length && $this->isDigit($this->sql[$this->pos])) {
                    $this->pos++;
                }
            }
        }

        // Handle scientific notation (e.g. 1.5e10, 1e-3, 2.5E+8)
        if ($this->pos < $this->length) {
            $c = $this->sql[$this->pos];
            if ($c === 'e' || $c === 'E') {
                $nextIdx = $this->pos + 1;
                if ($nextIdx < $this->length && ($this->sql[$nextIdx] === '+' || $this->sql[$nextIdx] === '-')) {
                    $nextIdx++;
                }
                if ($nextIdx < $this->length && $this->isDigit($this->sql[$nextIdx])) {
                    $isFloat = true;
                    $this->pos = $nextIdx;
                    while ($this->pos < $this->length && $this->isDigit($this->sql[$this->pos])) {
                        $this->pos++;
                    }
                }
            }
        }

        $value = substr($this->sql, $start, $this->pos - $start);
        $type = $isFloat ? TokenType::Float : TokenType::Integer;

        return new Token($type, $value, $start);
    }

    private function readIdentifierOrKeyword(int $start): Token
    {
        while ($this->pos < $this->length && $this->isIdentChar($this->sql[$this->pos])) {
            $this->pos++;
        }

        $value = substr($this->sql, $start, $this->pos - $start);
        $upper = strtoupper($value);

        if ($upper === 'NULL') {
            return new Token(TokenType::Null, $upper, $start);
        }

        if ($upper === 'TRUE' || $upper === 'FALSE') {
            return new Token(TokenType::Boolean, $upper, $start);
        }

        if (isset(self::KEYWORD_MAP[$upper])) {
            return new Token(TokenType::Keyword, $upper, $start);
        }

        return new Token(TokenType::Identifier, $value, $start);
    }

    private function readNamedPlaceholder(int $start): Token
    {
        $this->pos++;
        while ($this->pos < $this->length && $this->isIdentChar($this->sql[$this->pos])) {
            $this->pos++;
        }

        return new Token(TokenType::NamedPlaceholder, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function readNumberedPlaceholder(int $start): Token
    {
        $this->pos++;
        while ($this->pos < $this->length && $this->isDigit($this->sql[$this->pos])) {
            $this->pos++;
        }

        return new Token(TokenType::NumberedPlaceholder, substr($this->sql, $start, $this->pos - $start), $start);
    }

    private function tryReadOperator(int $start): ?Token
    {
        $char = $this->sql[$this->pos];
        $next = $this->peek(1);

        $twoChar = match (true) {
            $char === '<' && $next === '=' => '<=',
            $char === '>' && $next === '=' => '>=',
            $char === '!' && $next === '=' => '!=',
            $char === '<' && $next === '>' => '<>',
            $char === '|' && $next === '|' => '||',
            default => null,
        };

        if ($twoChar !== null) {
            $this->pos += 2;
            return new Token(TokenType::Operator, $twoChar, $start);
        }

        if (isset(self::SINGLE_OPERATORS[$char])) {
            $this->pos++;
            return new Token(TokenType::Operator, $char, $start);
        }

        return null;
    }
}
